File: problems/array/0053-maximum-subarray.js
/**
 * #53 Maximum Subarray
 * Pattern: Kadane's algorithm.
 * Invariant: bestEndingHere is the best sum of a subarray ending at this index.
 * Time: O(n) | Space: O(1)
 *
 * @param {number[]} nums
 * @return {number}
 */
function maxSubArray(nums) {
  let bestEndingHere = nums[0]
  let bestOverall = nums[0]

  for (let i = 1; i < nums.length; i += 1) {
    bestEndingHere = Math.max(nums[i], bestEndingHere + nums[i])
    bestOverall = Math.max(bestOverall, bestEndingHere)
  }

  return bestOverall
}

module.exports = maxSubArray

File: problems/array/0053-maximum-subarray.py
"""#53 Maximum Subarray
Kadane invariant: best_ending_here is the best sum ending at this index.
Time: O(n) | Space: O(1)
"""


class Solution:
    def maxSubArray(self, nums):
        best_ending_here = best_overall = nums[0]
        for number in nums[1:]:
            best_ending_here = max(number, best_ending_here + number)
            best_overall = max(best_overall, best_ending_here)
        return best_overall
